### ISSUE TITLE
The new 'Any' type and JavaScript

### ISSUE THREAD

**Issue by cristianmad** [google/protobuf#2612]
Hi, 

I recently started using protobuf for JavaScript. I have been using protobuf for the past few years but mostly with C#/C++.
Everything seems to go well, except I was not able to find a way to work with the new 'Any' type (defined in protobuf3). I am using the binary encoding which according to the doc, should be supported in JavaScript.

I have used protoc to define the JavaScript equivalent of my proto files. I can make instances of all my types, set properties, encode them (binary), etc. 

Basically, I do not seem to be able to access (see them) the Pack/Unpack functions.

Would anyone know how to do that? Any examples?

PS: I should add that my JavaScript expertise is very basic so the solution may be something obvious :)

Thanks!

**Comment by pherl** [comment]
@haberman

**Comment by cristianmad** [comment]
I figured it out! :)

Basically the JavaScript implementation of Any does not seem to have the Pack/Unpack functions defined. Instead I manually made instances of the Any type (where required) and then called the setType and setValue functions.

**proto**:
message Message
{
      Any AnyType = 0;
}

**JavaScript**

var message = new Message();
var value = new proto.google.protobuf.Any;
value.setValue(objectToPack.serializeBinary());
value.setTypeUrl("type.googleapis.com/FullyQualifedTypeNameOfTheObjectBeingPacked");

message.setAnyType(value);



There may be a better way of achieving this, but this worked for me.

Enjoy!

**Comment by haberman** [comment]
JavaScript does have `pack` and `unpack` -- see: https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/js/well_known_types/any.js

**Comment by cristianmad** [comment]
Thanks @haberman, 

I think that there may be an issue with my generated code since it does not contain the pack and unpack functions. It does have the other ones though (set/getType and set/getValue).

From the source code that you linked I see that the implementation is very similar to my workaround.

I tried calling pack on one of my Any instances and I got:
"0x800a01b6 - JavaScript runtime error: Object doesn't support property or method 'pack' occurred"

I think that something may have gone wrong when my JS protobuf code was generated.

Thank you

**Comment by cristianmad** [comment]
Also, note that when generating the code, I have followed the instructions documented here for CommonJS:
https://github.com/google/protobuf/tree/master/js

$ protoc --js_out=import_style=commonjs,binary:. messages.proto base.proto

The resulting file does not seem to contain the pack/unpack functions...(unless I am doing something wrong)

**Comment by haberman** [comment]
I should have mentioned that the well-known type support is not currently released (it's only on GitHub) but will be part of the upcoming 3.2 release.

**Comment by cristianmad** [comment]
Aha! :)

OK, thank you for your time! I will use my workaround until then.

**Comment by haberman** [comment]
This was released as part of 3.2! Resolving this issue.

**Comment by AlonikiX** [comment]
I don't know how to unpack an Any instance, could you please provide any example?

**Comment by psaelango** [comment]
@cristianmad  @haberman 
Can you please kindly share the updated solution for using protobuf.Any in javascript.
